sphereMove/SphereMove.cs

```diff
@@ -152,56 +200,115 @@ IEnumerator DelayedUpdate()
         // Run through interpolatePosX elements and update when the respective time is reached
         while (true)
         {
-            if (!new_trajectory)
+            switch (currentState)
             {
-                for (int i = 0; i < interpolatedPosX.Count; i++)
-                {
-                    car.x = interpolatedPosX[i];
-                    car.y = interpolatedPosY[i];
-                    car.date = interpolatedPosTime[i];
-                    Vector3 newPosition = new Vector3(car.x, car.y, this.gameObject.transform.localPosition.z);
-                    this.gameObject.transform.localPosition = newPosition;
-                    Debug.Log("Update x=" + this.gameObject.transform.localPosition.x.ToString() + "y=" + this.gameObject.transform.localPosition.y.ToString());
-
-                    // Wait until the next time to update the position
-                    yield return new WaitForSeconds((float)(interpolatedPosTime[i + 1] - interpolatedPosTime[i]).TotalSeconds);
-
-                }
-
-                // All interpolated positions have been used. Calculate new trajectory in the other coroutine
-                new_trajectory = true;
+                case State.Begin:
+                    if (!need_new_trajectory)
+                    {
+                        // TODO, if at the beginning stage the car is still, will give a NaN traj and give an error, ut should be fine bcs it changes to the Running stage quickly
+                        current_tracking_time = listTime[0];
+                        for (int i = 0; i < interpolatedPosX.Count; i++)
+                        {
+                            Thread.Sleep((int) Math.Round((interpolatedPosTime[i] - current_tracking_time).TotalSeconds * 1000)); // Blocked sleep in ms
+                            car.x = interpolatedPosX[i];
+                            car.y = interpolatedPosY[i];
+                            car.date = interpolatedPosTime[i].ToString("yyyy-MM-ddTHH:mm:ss.ffffff");
+                            Vector3 newPosition = new Vector3(car.x, car.y, this.gameObject.transform.localPosition.z);
+                            this.gameObject.transform.localPosition = newPosition;
+                            Debug.Log("Update x=" + this.gameObject.transform.localPosition.x.ToString() + "y=" + this.gameObject.transform.localPosition.y.ToString());
+                            current_tracking_time = interpolatedPosTime[i];
+                        }
+                        interpolatedPosTime.Clear();
+                        interpolatedPosX.Clear();
+                        interpolatedPosY.Clear();
+                        need_new_trajectory = true;
+                        currentState = State.Running;

```

Comment:
Updating the `currentState` will also impact the GetCarData function correct?

Reply:
Yes, it's a global flag to schedule both routine. State.Begin -> GetCarData plans the beginning traj -> Update runs the traj -> State.Running -> GetCarData plans the following trajectories -> ...